File: internal/ocr/biedronka.go
package ocr

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math"
	"regexp"
	"strconv"
	"strings"

	"github.com/shopspring/decimal"
)

// Tx is the loyalty-list metadata for one Biedronka purchase.
type Tx struct {
	ID         string
	Date       string
	StoreName  string
	ReceiptNum string
	TotalPrice float64
}

var shopNumber = regexp.MustCompile(`\d{3,5}`)

// BillFromBiedronka maps a fiscal e-receipt (or loyalty details) to a Bulkly bill.
func BillFromBiedronka(raw []byte, tx Tx) (Bill, error) {
	lines := extractBiedronkaLines(raw)
	if len(lines) == 0 {
		return Bill{}, ErrNoLines
	}
	scale := moneyScale(lines, tx.TotalPrice)
	var bill Bill
	bill.BoughtOn = strings.TrimSpace(tx.Date)
	bill.StoryName = strings.TrimSpace(tx.StoreName)
	if bill.StoryName == "" {
		bill.StoryName = "Biedronka"
	}
	if n := shopNumber.FindString(bill.StoryName); n != "" {
		bill.ExternalID = n
	}
	if num := strings.TrimSpace(tx.ReceiptNum); num != "" {
		bill.Notes = "Receipt " + num
	}
	var current int
	current = -1
	for _, item := range lines {
		if sell, ok := item["sellLine"].(map[string]any); ok {
			if truthy(sell["isStorno"]) {
				current = -1
				continue
			}
			line := Line{
				ReceiptName: strings.TrimSpace(asString(sell["name"])),
				Quantity:    formatQty(asFloat(sell["quantity"])),
				UnitPrice:   formatMoney(asFloat(sell["price"]) / scale),
				Amount:      formatMoney(asFloat(sell["total"]) / scale),
				VatType:     asString(sell["vatId"]),
			}
			if line.ReceiptName == "" {
				continue
			}
			bill.Lines = append(bill.Lines, line)
			current = len(bill.Lines) - 1
			continue
		}
		if disc, ok := item["discountLine"].(map[string]any); ok {
			if current < 0 || truthy(disc["isPercent"]) || truthy(disc["isStorno"]) {
				continue
			}
			add := asFloat(disc["value"]) / scale
			if add == 0 {
				continue
			}
			sum := asFloat(bill.Lines[current].Discount) + math.Abs(add)
			bill.Lines[current].Discount = formatMoney(sum)
		}
	}
	if len(bill.ProductLines()) == 0 {
		return Bill{}, ErrNoLines
	}
	payload, err := json.Marshal(bill)
	if err != nil {
		return Bill{}, err
	}
	return parseBill(payload)
}

func extractBiedronkaLines(raw []byte) []map[string]any {
	var payload any
	raw = bytes.TrimSpace(raw)
	if len(raw) == 0 {
		return nil
	}
	if err := json.Unmarshal(raw, &payload); err != nil {
		return nil
	}
	return walkBiedronkaLines(payload)
}

func walkBiedronkaLines(payload any) []map[string]any {
	switch v := payload.(type) {
	case []any:
		if looksLikeTillLines(v) {
			return asObjectSlice(v)
		}
		for _, item := range v {
			if found := walkBiedronkaLines(item); len(found) > 0 {
				return found
			}
		}
	case map[string]any:
		for _, key := range []string{"lines", "receiptLines"} {
			if arr, ok := v[key].([]any); ok {
				return asObjectSlice(arr)
			}
		}
		for _, key := range []string{"receipt", "data", "payload", "json"} {
			if found := walkBiedronkaLines(v[key]); len(found) > 0 {
				return found
			}
		}
		if items, ok := v["items"].([]any); ok {
			return linesFromDetails(items)
		}
	}
	return nil
}

func looksLikeTillLines(items []any) bool {
	for _, item := range items {
		obj, ok := item.(map[string]any)
		if !ok {
			continue
		}
		if _, ok := obj["sellLine"]; ok {
			return true
		}
		if _, ok := obj["discountLine"]; ok {
			return true
		}
		if _, ok := obj["sumInCurrency"]; ok {
			return true
		}
	}
	return false
}

func asObjectSlice(items []any) []map[string]any {
	out := make([]map[string]any, 0, len(items))
	for _, item := range items {
		if obj, ok := item.(map[string]any); ok {
			out = append(out, obj)
		}
	}
	return out
}

func linesFromDetails(items []any) []map[string]any {
	out := make([]map[string]any, 0, len(items))
	for _, item := range items {
		obj, ok := item.(map[string]any)
		if !ok {
			continue
		}
		sell := map[string]any{}
		if name := asString(obj["name"]); name != "" {
			sell["name"] = name
		}
		if _, ok := obj["price"]; ok {
			sell["price"] = obj["price"]
		} else if _, ok := obj["unit_price"]; ok {
			sell["price"] = obj["unit_price"]
		}
		if _, ok := obj["total"]; ok {
			sell["total"] = obj["total"]
		} else if _, ok := obj["total_price"]; ok {
			sell["total"] = obj["total_price"]
		}
		if _, ok := obj["quantity"]; ok {
			sell["quantity"] = obj["quantity"]
		}
		if _, ok := obj["vatId"]; ok {
			sell["vatId"] = obj["vatId"]
		}
		if len(sell) == 0 {
			continue
		}
		out = append(out, map[string]any{"sellLine": sell})
	}
	return out
}

func moneyScale(lines []map[string]any, listTotal float64) float64 {
	var sum float64
	for _, item := range lines {
		sell, ok := item["sellLine"].(map[string]any)
		if !ok || truthy(sell["isStorno"]) {
			continue
		}
		sum += asFloat(sell["total"])
	}
	if listTotal > 0 && sum > 0 {
		ratio := sum / listTotal
		if ratio > 50 && ratio < 150 {
			return 100
		}
	}
	return 1
}

func asFloat(v any) float64 {
	switch t := v.(type) {
	case float64:
		return t
	case float32:
		return float64(t)
	case int:
		return float64(t)
	case int64:
		return float64(t)
	case json.Number:
		n, _ := t.Float64()
		return n
	case string:
		s := strings.ReplaceAll(strings.TrimSpace(t), ",", ".")
		n, _ := strconv.ParseFloat(s, 64)
		return n
	case bool:
		return 0
	default:
		return 0
	}
}

func asString(v any) string {
	switch t := v.(type) {
	case string:
		return strings.TrimSpace(t)
	case json.Number:
		return t.String()
	case float64:
		if t == math.Trunc(t) {
			return strconv.FormatInt(int64(t), 10)
		}
		return strconv.FormatFloat(t, 'f', -1, 64)
	case int:
		return strconv.Itoa(t)
	case int64:
		return strconv.FormatInt(t, 10)
	default:
		return ""
	}
}

func truthy(v any) bool {
	switch t := v.(type) {
	case bool:
		return t
	case string:
		s := strings.ToLower(strings.TrimSpace(t))
		return s == "true" || s == "1"
	case float64:
		return t != 0
	default:
		return false
	}
}

func formatMoney(v float64) string {
	if math.IsNaN(v) || math.IsInf(v, 0) {
		return ""
	}
	return decimal.NewFromFloat(v).Round(2).StringFixed(2)
}

func formatQty(v float64) string {
	if math.IsNaN(v) || math.IsInf(v, 0) || v == 0 {
		return ""
	}
	d := decimal.NewFromFloat(v)
	if d.Equal(d.Truncate(0)) {
		return d.Truncate(0).String()
	}
	return d.String()
}

func BillSlipText(bill Bill, tx Tx) string {
	var b strings.Builder
	b.WriteString("Biedronka e-receipt\n")
	if tx.StoreName != "" {
		b.WriteString(tx.StoreName)
		b.WriteByte('\n')
	}
	if tx.Date != "" {
		b.WriteString(tx.Date)
		b.WriteByte('\n')
	}
	if tx.ReceiptNum != "" {
		fmt.Fprintf(&b, "Receipt %s\n", tx.ReceiptNum)
	}
	b.WriteByte('\n')
	for _, line := range bill.ProductLines() {
		name := line.ReceiptName
		if name == "" {
			name = line.ProductName
		}
		fmt.Fprintf(&b, "%s  %s x %s  %s\n", name, line.Quantity, line.UnitPrice, line.Amount)
	}
	return b.String()
}

File: src/ocr/biedronka.ts
import Decimal from 'decimal.js';
import { parseBill } from './parse';
import { emptyBill, NoLinesError, productLines, type Bill, type Line } from './types';

export type Tx = {
  id: string;
  Date: string;
  StoreName: string;
  ReceiptNum: string;
  TotalPrice: number;
};

const shopNumber = /\d{3,5}/;

export function billFromBiedronka(raw: Buffer | string, tx: Tx): Bill {
  const lines = extractBiedronkaLines(typeof raw === 'string' ? Buffer.from(raw) : raw);
  if (lines.length === 0) throw new NoLinesError();
  const scale = moneyScale(lines, tx.TotalPrice);
  const bill = emptyBill();
  bill.boughtOn = tx.Date.trim();
  bill.storyName = tx.StoreName.trim();
  if (bill.storyName === '') bill.storyName = 'Biedronka';
  const n = shopNumber.exec(bill.storyName);
  if (n) bill.externalId = n[0]!;
  const num = tx.ReceiptNum.trim();
  if (num !== '') bill.notes = 'Receipt ' + num;
  let current = -1;
  for (const item of lines) {
    const sell = asObject(item.sellLine);
    if (sell) {
      if (truthy(sell.isStorno)) {
        current = -1;
        continue;
      }
      const line: Line = {
        receiptName: asString(sell.name).trim(),
        productName: '',
        productId: 0,
        unitId: 0,
        unitName: '',
        vatType: asString(sell.vatId),
        packageCount: '',
        packageSize: '',
        quantity: formatQty(asFloat(sell.quantity)),
        unitPrice: formatMoney(asFloat(sell.price) / scale),
        discount: '',
        amount: formatMoney(asFloat(sell.total) / scale),
        skip: false,
        skipReason: '',
      };
      if (line.receiptName === '') continue;
      bill.lines.push(line);
      current = bill.lines.length - 1;
      continue;
    }
    const disc = asObject(item.discountLine);
    if (disc) {
      if (current < 0 || truthy(disc.isPercent) || truthy(disc.isStorno)) continue;
      const add = asFloat(disc.value) / scale;
      if (add === 0) continue;
      const sum = asFloat(bill.lines[current]!.discount) + Math.abs(add);
      bill.lines[current]!.discount = formatMoney(sum);
    }
  }
  if (productLines(bill).length === 0) throw new NoLinesError();
  return parseBill(Buffer.from(JSON.stringify(toLooseJSON(bill))));
}

function toLooseJSON(bill: Bill): unknown {
  return {
    bought_on: bill.boughtOn,
    bought_at: bill.boughtAt,
    notes: bill.notes,
    not_a_bill: bill.notABill,
    company_id: bill.storyId,
    company_name: bill.storyName,
    external_id: bill.externalId,
    street_name: bill.streetName,
    building_number: bill.buildingNumber,
    apartment_number: bill.apartmentNumber,
    postal_code: bill.postalCode,
    city: bill.city,
    lines: bill.lines.map((line) => ({
      receipt_name: line.receiptName,
      product_name: line.productName,
      product_id: line.productId,
      unit_id: line.unitId,
      unit_name: line.unitName,
      vat_type: line.vatType,
      package_count: line.packageCount,
      package_size: line.packageSize,
      quantity: line.quantity,
      unit_price: line.unitPrice,
      discount: line.discount,
      amount: line.amount,
      skip: line.skip,
      skip_reason: line.skipReason,
    })),
  };
}

function extractBiedronkaLines(raw: Buffer): Record<string, unknown>[] {
  const text = raw.toString('utf8').trim();
  if (text === '') return [];
  let payload: unknown;
  try {
    payload = JSON.parse(text);
  } catch {
    return [];
  }
  return walkBiedronkaLines(payload);
}

function walkBiedronkaLines(payload: unknown): Record<string, unknown>[] {
  if (Array.isArray(payload)) {
    if (looksLikeTillLines(payload)) return asObjectSlice(payload);
    for (const item of payload) {
      const found = walkBiedronkaLines(item);
      if (found.length > 0) return found;
    }
  } else if (payload && typeof payload === 'object') {
    const v = payload as Record<string, unknown>;
    for (const key of ['lines', 'receiptLines']) {
      if (Array.isArray(v[key])) return asObjectSlice(v[key]);
    }
    for (const key of ['receipt', 'data', 'payload', 'json']) {
      const found = walkBiedronkaLines(v[key]);
      if (found.length > 0) return found;
    }
    if (Array.isArray(v.items)) return linesFromDetails(v.items);
  }
  return [];
}

function looksLikeTillLines(items: unknown[]): boolean {
  for (const item of items) {
    if (!item || typeof item !== 'object' || Array.isArray(item)) continue;
    const obj = item as Record<string, unknown>;
    if ('sellLine' in obj || 'discountLine' in obj || 'sumInCurrency' in obj) return true;
  }
  return false;
}

function asObjectSlice(items: unknown[]): Record<string, unknown>[] {
  const out: Record<string, unknown>[] = [];
  for (const item of items) {
    if (item && typeof item === 'object' && !Array.isArray(item)) out.push(item as Record<string, unknown>);
  }
  return out;
}

function linesFromDetails(items: unknown[]): Record<string, unknown>[] {
  const out: Record<string, unknown>[] = [];
  for (const item of items) {
    if (!item || typeof item !== 'object' || Array.isArray(item)) continue;
    const obj = item as Record<string, unknown>;
    const sell: Record<string, unknown> = {};
    const name = asString(obj.name);
    if (name !== '') sell.name = name;
    if ('price' in obj) sell.price = obj.price;
    else if ('unit_price' in obj) sell.price = obj.unit_price;
    if ('total' in obj) sell.total = obj.total;
    else if ('total_price' in obj) sell.total = obj.total_price;
    if ('quantity' in obj) sell.quantity = obj.quantity;
    if ('vatId' in obj) sell.vatId = obj.vatId;
    if (Object.keys(sell).length === 0) continue;
    out.push({ sellLine: sell });
  }
  return out;
}

function moneyScale(lines: Record<string, unknown>[], listTotal: number): number {
  let sum = 0;
  for (const item of lines) {
    const sell = asObject(item.sellLine);
    if (!sell || truthy(sell.isStorno)) continue;
    sum += asFloat(sell.total);
  }
  if (listTotal > 0 && sum > 0) {
    const ratio = sum / listTotal;
    if (ratio > 50 && ratio < 150) return 100;
  }
  return 1;
}

function asObject(v: unknown): Record<string, unknown> | null {
  if (v && typeof v === 'object' && !Array.isArray(v)) return v as Record<string, unknown>;
  return null;
}

function asFloat(v: unknown): number {
  if (typeof v === 'number') return Number.isFinite(v) ? v : 0;
  if (typeof v === 'string') {
    const n = Number.parseFloat(v.trim().replaceAll(',', '.'));
    return Number.isFinite(n) ? n : 0;
  }
  if (typeof v === 'boolean') return 0;
  return 0;
}

function asString(v: unknown): string {
  if (typeof v === 'string') return v.trim();
  if (typeof v === 'number') {
    if (v === Math.trunc(v)) return String(Math.trunc(v));
    return String(v);
  }
  return '';
}

function truthy(v: unknown): boolean {
  if (typeof v === 'boolean') return v;
  if (typeof v === 'string') {
    const s = v.toLowerCase().trim();
    return s === 'true' || s === '1';
  }
  if (typeof v === 'number') return v !== 0;
  return false;
}

function formatMoney(v: number): string {
  if (!Number.isFinite(v)) return '';
  return new Decimal(v).toDecimalPlaces(2, Decimal.ROUND_HALF_EVEN).toFixed(2);
}

function formatQty(v: number): string {
  if (!Number.isFinite(v) || v === 0) return '';
  const d = new Decimal(v);
  if (d.equals(d.trunc())) return d.trunc().toString();
  return d.toString();
}

export function billSlipText(bill: Bill, tx: Tx): string {
  let out = 'Biedronka e-receipt\n';
  if (tx.StoreName !== '') out += tx.StoreName + '\n';
  if (tx.Date !== '') out += tx.Date + '\n';
  if (tx.ReceiptNum !== '') out += `Receipt ${tx.ReceiptNum}\n`;
  out += '\n';
  for (const line of productLines(bill)) {
    let name = line.receiptName;
    if (name === '') name = line.productName;
    out += `${name}  ${line.quantity} x ${line.unitPrice}  ${line.amount}\n`;
  }
  return out;
}
